src/main/scala/com/zeotap/data/io/source/spark/constructs/DataFrameOps.scala

```diff
@@ -71,6 +76,52 @@ object DataFrameOps {
 
       dataFrame.withColumn("inputPathColumn", addInputPathColumn(input_file_name())).select("inputPathColumn").distinct().collect.map(row => row.getString(0))
     }
+
+    /**
+     * Partitions given dataframe into number of partitions provided and writes the
+     * partitioned dataframe to the intermediate path provided, so that parallel processing can take place.
+     *
+     * @param numberOfPartitions         is the number of partitions we want in the dataframe.
+     * @param intermediatePath           is the intermediate path in which the partitioned data frame is written.
+     * @param prioritiseIntermediatePath is boolean value which denotes whether the intermediate path (i.e already partitioned path)
+     *                                   needs to prioritised or should we force repartition. If it is true then data at intermediate path
+     *                                   will be returned(if non empty).
+     * @return Returns Dataframe with specified number of partitions.
+     */
+
+    def split(numberOfPartitions: Int, intermediatePath: String, prioritiseIntermediatePath: Boolean)(implicit spark: SparkSession): DataFrame = {
+      val rawInputDf = dataFrame
+      if (!prioritiseIntermediatePath) {
+        defaultSplit(rawInputDf, numberOfPartitions, intermediatePath)
+      }
+      else {
+        val intermediateDf = util.Try {
+          val df: DataFrame = ParquetSparkLoader().load(intermediatePath).buildUnsafe(spark)
+          if (df.isEmpty) throw new IllegalArgumentException("Intermediate Data at " + intermediatePath + " is empty!")
+          else df
+        }
+
+        intermediateDf match {
+          case Success(df) => df
+          case Failure(exception) =>
+            logger.log.info("Warning : " + exception.getMessage + "\ncontinuing with default splitting strategy!")
+            defaultSplit(rawInputDf, numberOfPartitions, intermediatePath)
+        }
+      }
+    }
+
+    /*
+    Default Splitting strategy, takes Raw Input Dataframe , re-partitions it and returns it for further processing.
+     */
+    def defaultSplit(rawInputDf: DataFrame, numberOfPartitions: Int, intermediatePath: String)(implicit spark: SparkSession): DataFrame = {
+      val partitionedDf = rawInputDf.repartition(numberOfPartitions)
+      ParquetSparkWriter().addSaveMode(Overwrite).save(intermediatePath).buildUnsafe(partitionedDf)
+      ParquetSparkLoader().load(intermediatePath).buildUnsafe(spark)
+    }
   }
 
 }
+
+object logger extends Serializable {
+  @transient lazy val log: Logger = Logger.getLogger(this.getClass.getName)
+}
```

Comment:
move this to util in a separate class so that it is reused across project. make sure it is a Singleton